### DISCUSSION TITLE
如何让 xmake 自动调用外部工具生成源码，并把源码加入项目中？

### DISCUSSION THREAD

**Discussion by vrecluse** [xmake-io/xmake#3795] [Q&A]
比如有个外部工具，通过一些文件，在 temp 目录下生成一些 C++ 代码
生成逻辑比较复杂，源文件和 C++ 代码没有简单的一一对应关系，所以也比较难定义 xmake 的 rule 来生成
现在希望通过 xmake 在编译时，自动先调用代码生成工具，再把 temp 目录下的代码加入到 项目中
如
```
target("code_generation")
   on_build(function (target) 
      -- generate code...
   end
   )

target("test")
  add_deps("code_generation")
  add_files("temp/**.cpp")
```

但这样 add_files 似乎找不到生成的 cpp 文件。

如果把生成代码的调用放到 on_load 执行，然后在 after_load 里去调用 target:add("fle", "temp/**.cpp")

似乎能解决问题，但依赖的 target 多了也有点麻烦，并且运行 xmake 的其它命令，如 run 等，也会出发 on_load 重新生成代码。

正确的姿势应该是什么？

**Comment by waruqi** [comment]
https://github.com/xmake-io/xmake/blob/master/tests/projects/other/autogen_codedep/xmake.lua

**Reply by vrecluse** [reply]
问题的场景跟示例还不太一样
源文件生成 cpp 的规则比较复杂，无法直接通过文件名就知道生成会什么 cpp 文件
比如有源文件 a.in，有可能会生成 b.cpp c.cpp x.cpp 一个或多个文件，也有可能不生成
自定义 rule 似乎不太好弄

**Reply by vrecluse** [reply]
基本上需要把外部工具的批量生成逻辑当作黑箱处理，只能在它生成完cpp代码后加进target

**Reply by waruqi** [reply]
没啥不好弄的，又没规定 rule 只能一个 源文件输出一个 cpp

https://github.com/xmake-io/xmake/blob/eeb6658c829a0e2d72d47790b98a6669a74c444e/xmake/rules/asn1c/xmake.lua#L36

输出多个不也可以么。。

**Reply by vrecluse** [reply]
这样会不会导致同目录下的其它 .c 文件被重复尝试编译多次？
比如 a.in 先生成了 x.c y.c z.c
b.in 生成时又把之前生成的 x.c y.c z.c 给添加了一遍

**Reply by waruqi** [reply]
这就是你自己 rule 脚本里面处理了。。上面我只是给你个 demo 。。不要照搬。。
